crates/bevy_ecs/src/bundle/info.rs

```diff
@@ -111,41 +112,106 @@ impl BundleInfo {
             panic!("Bundle {bundle_type_name} has duplicate components: {names:?}");
         }
 
-        // handle explicit components
         let explicit_components_len = component_ids.len();
+
+        // SAFETY: The caller ensures that component_ids:
+        // - is valid for the associated world
+        // - has had its storage initialized
+        // - is in the same order as the source bundle type
+        let mut info = BundleInfo {
+            id,
+            component_ids,
+            required_components: Vec::new(),
+            explicit_components_len,
+        };
+
+        // SAFETY: first call with the same `storages` and `components`.
+        unsafe {
+            info.set_required_components(storages, components, components_in_bundles, |_| true);
+        }
+
+        info
+    }
+
+    /// Update [`Self::component_ids`] to contain all required components, updates [`Self::required_components`] alongside.
+    ///
+    /// The filter `components_in_bundles_filter` determines for which components [`Self::id`] is added to `components_in_bundles`.
+    ///
+    /// # Safety
+    ///
+    /// `Self` must be constructed from [`Self::new`] with the same `storages` and `components`.
+    unsafe fn set_required_components(
+        &mut self,
+        storages: &mut Storages,
+        components: &Components,
+        components_in_bundles: &mut Vec<Vec<BundleId>>,
+        components_in_bundles_filter: impl Fn(ComponentId) -> bool,
+    ) {
+        let mut components_in_bundles_push = |component: ComponentId| {
+            if !components_in_bundles_filter(component) {
+                return;
+            }
+            match components_in_bundles.get_mut(component.index()) {
+                Some(bundles) => bundles.push(self.id),
+                None => {
+                    let missing_minus_one = component.index() - components_in_bundles.len();
+                    let iter = core::iter::repeat_n(Vec::new(), missing_minus_one)
+                        .chain(core::iter::once(vec![self.id]));
+                    components_in_bundles.reserve_exact(missing_minus_one + 1);
+                    components_in_bundles.extend(iter);
```

Comment:
Could this be simplified with a ` resize_with`? I'm not sure iterators will optimize nicely here, and it would also feel more readable IMO.